File: tara-web/src/lib/types/cra.ts
/**
 * CRA Compliance TypeScript types
 *
 * Used by: lib/api/craApi.ts, features/cra/*, routes/cra/*
 */

export type CraClassification = 'default' | 'class_i' | 'class_ii' | 'critical';

export type CraProductType = 'current' | 'legacy_a' | 'legacy_b' | 'legacy_c';

export type CraAssessmentStatus = 'draft' | 'in_progress' | 'complete';

export type CraRequirementStatus = 'not_started' | 'partial' | 'compliant' | 'not_applicable';

export type CompensatingControlStatus = 'planned' | 'implemented' | 'verified';

export type GapSeverity = 'none' | 'low' | 'medium' | 'high' | 'critical';

export type CompliancePath = 'direct_patch' | 'compensating_control' | 'hybrid';

export type TargetMarket = 'eu' | 'non_eu' | 'global';

export interface InventoryItem {
  readonly id: string;
  readonly assessment_id: string;
  readonly sku: string;
  readonly firmware_version?: string;
  readonly units_in_stock: number;
  readonly units_in_field: number;
  readonly oem_customer?: string;
  readonly target_market: TargetMarket;
  readonly last_production_date?: string;
  readonly notes?: string;
  readonly created_at?: string;
  readonly updated_at?: string;
}

export interface InventorySummary {
  readonly total_skus: number;
  readonly total_units_in_stock: number;
  readonly total_units_in_field: number;
  readonly eu_units: number;
  readonly non_eu_units: number;
  readonly oems: string[];
}

export interface DataProfileQuestion {
  readonly key: string;
  readonly label: string;
  readonly help_text: string;
  readonly category: string;
}

export interface DataProfile {
  readonly [key: string]: boolean;
}

export interface ApplicabilityResult {
  readonly requirement_id: string;
  readonly applicable: boolean;
  readonly justification: string;
}

export interface DataProfileResponse {
  readonly profile: DataProfile;
  readonly applicability: readonly ApplicabilityResult[];
  readonly auto_resolved_count: number;
}

export interface ClassificationQuestion {
  readonly id: string;
  readonly text: string;
  readonly hint: string;
}

export interface ConformityModule {
  readonly module_id: string;
  readonly name: string;
  readonly description: string;
  readonly mandatory: boolean;
  readonly alternatives: readonly string[];
  readonly rationale: string;
}

export interface ClassificationResult {
  readonly classification: CraClassification;
  readonly category_id?: string;
  readonly category_name: string;
  readonly conformity_assessment: string;
  readonly conformity_module: ConformityModule;
  readonly compliance_deadline: string;
  readonly reporting_deadline: string;
  readonly cost_estimate_min: number;
  readonly cost_estimate_max: number;
  readonly automotive_exception: boolean;
  readonly rationale: string;
}

export interface ProductCategory {
  readonly id: string;
  readonly name: string;
  readonly classification: CraClassification;
  readonly description: string;
  readonly examples: string;
  readonly annex_ref: string;
}

export type AnnexPart = 'Part I' | 'Part II' | 'Documentation';
export type ObligationType = 'risk_based' | 'mandatory';

export interface CraRequirementDefinition {
  readonly id: string;
  readonly name: string;
  readonly article: string;
  readonly category: string;
  readonly annex_part: AnnexPart;
  readonly obligation_type: ObligationType;
}

export interface SubRequirement {
  readonly description: string;
  readonly check_evidence: string;
  readonly typical_gap: string;
}

export interface RemediationAction {
  readonly action: string;
  readonly owner_hint: string;
  readonly effort_days: number;
}

export interface RequirementGuidance {
  readonly requirement_id: string;
  readonly annex_section: string;
  readonly cra_article: string;
  readonly priority: string;
  readonly deadline_note: string;
  readonly explanation: string;
  readonly regulatory_text: string;
  readonly sub_requirements: readonly SubRequirement[];
  readonly evidence_checklist: readonly string[];
  readonly investigation_prompts: readonly string[];
  readonly common_gaps: readonly string[];
  readonly remediation_actions: readonly RemediationAction[];
  readonly effort_estimate: string;
  readonly mapped_controls: readonly string[];
  readonly mapped_standards: readonly string[];
  readonly tara_link: string;
}

export interface CraRequirementStatusRecord {
  readonly id: string;
  readonly assessment_id: string;
  readonly requirement_id: string;
  readonly requirement_name?: string;
  readonly requirement_article?: string;
  readonly requirement_category?: string;
  readonly status: CraRequirementStatus;
  readonly auto_mapped: boolean;
  readonly mapped_artifact_type?: string;
  readonly mapped_artifact_count: number;
  readonly owner?: string;
  readonly target_date?: string;
  readonly evidence_notes?: string;
  readonly evidence_links: string[];
  readonly gap_description?: string;
  readonly remediation_plan?: string;
  readonly gap_severity: GapSeverity;
  readonly residual_risk_level: GapSeverity;
  readonly created_at?: string;
  readonly updated_at?: string;
}

export interface MitigatedRequirementInfo {
  readonly requirement_status_id: string;
  readonly requirement_id: string;
  readonly requirement_name?: string;
}

export interface CompensatingControl {
  readonly id: string;
  readonly assessment_id: string;
  readonly control_id: string;
  readonly name: string;
  readonly description?: string;
  readonly implementation_status: CompensatingControlStatus;
  readonly supplier_actions?: string;
  readonly oem_actions?: string;
  readonly residual_risk?: string;
  readonly mitigated_requirements: MitigatedRequirementInfo[];
  readonly created_at?: string;
  readonly updated_at?: string;
}

export interface CraAssessment {
  readonly id: string;
  readonly product_id: string;
  readonly product_name?: string;
  readonly classification?: CraClassification;
  readonly classification_answers: Record<string, boolean>;
  readonly product_type: CraProductType;
  readonly compliance_path: CompliancePath;
  readonly compliance_deadline?: string;
  readonly assessment_date?: string;
  readonly assessor_id?: string;
  readonly status: CraAssessmentStatus;
  readonly overall_compliance_pct: number;
  readonly support_period_years?: number;
  readonly support_period_justification?: string;
  readonly support_period_end?: string;
  readonly eoss_date?: string;
  readonly notes?: string;
  readonly automotive_exception: boolean;
  readonly created_at?: string;
  readonly updated_at?: string;
  readonly requirement_statuses: CraRequirementStatusRecord[];
  readonly compensating_controls: CompensatingControl[];
}

export interface CraAssessmentListItem {
  readonly id: string;
  readonly product_id: string;
  readonly product_name?: string;
  readonly classification?: CraClassification;
  readonly product_type: CraProductType;
  readonly status: CraAssessmentStatus;
  readonly overall_compliance_pct: number;
  readonly compliance_deadline?: string;
  readonly updated_at?: string;
}

export interface CraAssessmentListResponse {
  readonly assessments: CraAssessmentListItem[];
  readonly total: number;
}

export interface CreateAssessmentRequest {
  readonly product_id: string;
  readonly product_type: CraProductType;
  readonly notes?: string;
}

export interface UpdateAssessmentRequest {
  readonly status?: CraAssessmentStatus;
  readonly product_type?: CraProductType;
  readonly support_period_years?: number;
  readonly support_period_justification?: string;
  readonly support_period_end?: string;
  readonly eoss_date?: string;
  readonly notes?: string;
}

export interface ClassifyRequest {
  readonly category_id?: string;
  readonly uses_harmonised_standard: boolean;
  readonly is_open_source_public: boolean;
  readonly automotive_exception: boolean;
  readonly answers?: Record<string, boolean>;
}

export interface UpdateRequirementRequest {
  readonly status?: CraRequirementStatus;
  readonly owner?: string;
  readonly target_date?: string;
  readonly evidence_notes?: string;
  readonly evidence_links?: string[];
  readonly gap_description?: string;
  readonly remediation_plan?: string;
  readonly gap_severity?: GapSeverity;
  readonly residual_risk_level?: GapSeverity;
}

export interface CreateCompensatingControlRequest {
  readonly control_id: string;
  readonly name: string;
  readonly description?: string;
  readonly implementation_status?: CompensatingControlStatus;
  readonly supplier_actions?: string;
  readonly oem_actions?: string;
  readonly residual_risk?: string;
  readonly mitigated_requirement_ids?: string[];
}

export interface UpdateCompensatingControlRequest {
  readonly name?: string;
  readonly description?: string;
  readonly implementation_status?: CompensatingControlStatus;
  readonly supplier_actions?: string;
  readonly oem_actions?: string;
  readonly residual_risk?: string;
  readonly mitigated_requirement_ids?: string[];
}

export interface AutoMapResponse {
  readonly mapped_count: number;
  readonly mappings: Array<{
    readonly requirement_id: string;
    readonly status: string;
    readonly artifact_type: string;
    readonly artifact_count: number;
    readonly evidence_notes: string;
  }>;
}

export interface ClassificationQuestionsResponse {
  readonly questions: ClassificationQuestion[];
  readonly automotive_exception_question: ClassificationQuestion;
}

export interface CompensatingControlCatalogItem {
  readonly control_id: string;
  readonly name: string;
  readonly category: string;
  readonly description: string;
  readonly applicability: string;
  readonly limitations: string;
  readonly applicable_requirements: string[];
}

export type RiskLevel = 'none' | 'low' | 'medium' | 'high' | 'critical';

export interface GapGuidance {
  readonly priority: string;
  readonly deadline_note: string;
  readonly effort_estimate: string;
  readonly cra_article: string;
  readonly explanation: string;
  readonly common_gaps: readonly string[];
  readonly sub_requirements: readonly SubRequirement[];
  readonly remediation_actions: readonly RemediationAction[];
  readonly mapped_standards: readonly string[];
}

export interface GapRequirementItem {
  readonly requirement_status_id: string;
  readonly requirement_id: string;
  readonly requirement_name: string;
  readonly category: string;
  readonly article: string;
  readonly annex_part: AnnexPart;
  readonly obligation_type: ObligationType;
  readonly status: CraRequirementStatus;
  readonly is_gap: boolean;
  readonly risk_level: RiskLevel;
  readonly suggested_controls: string[];
  readonly applied_controls: Array<{
    readonly control_id: string;
    readonly name: string;
    readonly status: string;
  }>;
  readonly tara_evidence: Array<{
    readonly type: string;
    readonly count: number;
    readonly notes: string;
  }>;
  readonly owner?: string;
  readonly target_date?: string;
  readonly evidence_notes?: string;
  readonly guidance?: GapGuidance;
}

export interface GapAnalysisSummary {
  readonly total: number;
  readonly compliant: number;
  readonly gaps: number;
  readonly critical_risk: number;
  readonly high_risk: number;
  readonly with_controls: number;
  readonly mitigated: number;
  readonly unmitigated: number;
  readonly total_remediation_effort_days: number;
  readonly risk_reduction_pct: number;
}

export interface GapAnalysisResponse {
  readonly assessment_id: string;
  readonly product_id: string;
  readonly product_type: CraProductType;
  readonly classification: CraClassification;
  readonly is_legacy: boolean;
  readonly requirements: GapRequirementItem[];
  readonly summary: GapAnalysisSummary;
}

File: api/models/cra.py
"""
CRA Compliance Pydantic models for request/response schemas.

Depends on: core.cra_classifier, core.cra_auto_mapper
Used by: api/routes/cra.py
"""
from enum import Enum
from typing import List, Optional, Dict, Any
from datetime import datetime
from pydantic import BaseModel, Field


class CraClassificationEnum(str, Enum):
    """CRA product classification"""
    DEFAULT = "default"
    CLASS_I = "class_i"
    CLASS_II = "class_ii"
    CRITICAL = "critical"


class CraProductTypeEnum(str, Enum):
    """Product lifecycle type"""
    CURRENT = "current"
    LEGACY_A = "legacy_a"
    LEGACY_B = "legacy_b"
    LEGACY_C = "legacy_c"


class CraAssessmentStatusEnum(str, Enum):
    """Assessment workflow status"""
    DRAFT = "draft"
    IN_PROGRESS = "in_progress"
    COMPLETE = "complete"


class CraRequirementStatusEnum(str, Enum):
    """Requirement compliance status"""
    NOT_STARTED = "not_started"
    PARTIAL = "partial"
    COMPLIANT = "compliant"
    NOT_APPLICABLE = "not_applicable"


class CompensatingControlStatusEnum(str, Enum):
    """Compensating control implementation status"""
    PLANNED = "planned"
    IMPLEMENTED = "implemented"
    VERIFIED = "verified"


class GapSeverityEnum(str, Enum):
    """Gap severity / residual risk level"""
    NONE = "none"
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class CompliancePathEnum(str, Enum):
    """CRA compliance path per Art. 5(3)"""
    DIRECT_PATCH = "direct_patch"
    COMPENSATING_CONTROL = "compensating_control"
    HYBRID = "hybrid"


# --- Request schemas ---

class CraAssessmentCreate(BaseModel):
    """Create a new CRA assessment for a product"""
    product_id: str = Field(..., description="Product scope ID")
    product_type: CraProductTypeEnum = Field(
        default=CraProductTypeEnum.CURRENT,
        description="Product lifecycle type"
    )
    notes: Optional[str] = Field(None, description="Assessment notes")


MIN_SUPPORT_PERIOD_YEARS: int = 5


class CraAssessmentUpdate(BaseModel):
    """Update an existing CRA assessment"""
    status: Optional[CraAssessmentStatusEnum] = None
    product_type: Optional[CraProductTypeEnum] = None
    support_period_years: Optional[int] = Field(
        None,
        description="Support period in years. CRA minimum is 5 years."
    )
    support_period_justification: Optional[str] = Field(
        None,
        description="Required if support_period_years < 5. Must justify shorter period."
    )
    support_period_end: Optional[str] = None
    eoss_date: Optional[str] = None
    notes: Optional[str] = None


class DataProfileUpdate(BaseModel):
    """Product data classification profile — boolean flags"""
    stores_data_at_rest: Optional[bool] = None
    stores_personal_data: Optional[bool] = None
    collects_telemetry: Optional[bool] = None
    has_network_interfaces: Optional[bool] = None
    has_user_authentication: Optional[bool] = None
    has_physical_interfaces: Optional[bool] = None
    has_updateable_software: Optional[bool] = None
    uses_third_party_components: Optional[bool] = None
    has_crypto_keys: Optional[bool] = None


class ApplicabilityResultResponse(BaseModel):
    """Whether a CRA requirement applies based on data profile"""
    requirement_id: str
    applicable: bool
    justification: str


class DataProfileResponse(BaseModel):
    """Full data profile with computed applicability"""
    profile: Dict[str, bool]
    applicability: List[ApplicabilityResultResponse]
    auto_resolved_count: int


class ClassifyRequest(BaseModel):
    """Classification based on core functionality category selection."""
    category_id: Optional[str] = Field(
        None,
        description="Product category ID from CRA Annexes III/IV catalog. None = Default."
    )
    uses_harmonised_standard: bool = Field(
        default=False,
        description="Whether a harmonised standard will be applied (affects Module A eligibility)"
    )
    is_open_source_public: bool = Field(
        default=False,
        description="Whether product is free/open-source with public technical docs"
    )
    automotive_exception: bool = Field(
        default=False,
        description="Whether UN R155 lex specialis applies"
    )
    answers: Dict[str, bool] = Field(
        default={},
        description="Legacy field — kept for backward compatibility"
    )


class RequirementStatusUpdate(BaseModel):
    """Update a single requirement status"""
    status: Optional[CraRequirementStatusEnum] = None
    owner: Optional[str] = None
    target_date: Optional[str] = None
    evidence_notes: Optional[str] = None
    evidence_links: Optional[List[str]] = None
    gap_description: Optional[str] = None
    remediation_plan: Optional[str] = None
    gap_severity: Optional[GapSeverityEnum] = None
    residual_risk_level: Optional[GapSeverityEnum] = None


class CompensatingControlCreate(BaseModel):
    """Create a compensating control"""
    control_id: str = Field(..., description="Control catalog ID")
    name: str = Field(..., description="Control name")
    description: Optional[str] = None
    implementation_status: CompensatingControlStatusEnum = Field(
        default=CompensatingControlStatusEnum.PLANNED
    )
    supplier_actions: Optional[str] = None
    oem_actions: Optional[str] = None
    residual_risk: Optional[str] = None
    mitigated_requirement_ids: Optional[List[str]] = Field(
        default=[],
        description="List of requirement_status IDs this control mitigates"
    )


class CompensatingControlUpdate(BaseModel):
    """Update a compensating control"""
    name: Optional[str] = None
    description: Optional[str] = None
    implementation_status: Optional[CompensatingControlStatusEnum] = None
    supplier_actions: Optional[str] = None
    oem_actions: Optional[str] = None
    residual_risk: Optional[str] = None
    mitigated_requirement_ids: Optional[List[str]] = None


# --- Response schemas ---

class CraRequirementDefinition(BaseModel):
    """Static definition of a CRA requirement"""
    id: str
    name: str
    article: str
    category: str
    annex_part: str = "Part I"
    obligation_type: str = "risk_based"


class SubRequirementResponse(BaseModel):
    """One checkable sub-item within a CRA requirement"""
    description: str
    check_evidence: str
    typical_gap: str


class RemediationActionResponse(BaseModel):
    """One concrete step to close a gap"""
    action: str
    owner_hint: str
    effort_days: int


class RequirementGuidanceResponse(BaseModel):
    """Per-requirement coaching data — coach AND doer"""
    requirement_id: str
    annex_section: str
    cra_article: str
    priority: str
    deadline_note: str
    explanation: str
    regulatory_text: str
    sub_requirements: List[SubRequirementResponse]
    evidence_checklist: List[str]
    investigation_prompts: List[str]
    common_gaps: List[str]
    remediation_actions: List[RemediationActionResponse]
    effort_estimate: str
    mapped_controls: List[str]
    mapped_standards: List[str]
    tara_link: str


class CraRequirementStatusResponse(BaseModel):
    """Requirement status within an assessment"""
    id: str
    assessment_id: str
    requirement_id: str
    requirement_name: Optional[str] = None
    requirement_article: Optional[str] = None
    requirement_category: Optional[str] = None
    status: str
    auto_mapped: bool
    mapped_artifact_type: Optional[str] = None
    mapped_artifact_count: int = 0
    owner: Optional[str] = None
    target_date: Optional[str] = None
    evidence_notes: Optional[str] = None
    evidence_links: List[str] = []
    gap_description: Optional[str] = None
    remediation_plan: Optional[str] = None
    gap_severity: str = "none"
    residual_risk_level: str = "none"
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class MitigatedRequirementInfo(BaseModel):
    """Brief info about a mitigated requirement"""
    requirement_status_id: str
    requirement_id: str
    requirement_name: Optional[str] = None


class CompensatingControlResponse(BaseModel):
    """Compensating control response"""
    id: str
    assessment_id: str
    control_id: str
    name: str
    description: Optional[str] = None
    implementation_status: str
    supplier_actions: Optional[str] = None
    oem_actions: Optional[str] = None
    residual_risk: Optional[str] = None
    mitigated_requirements: List[MitigatedRequirementInfo] = []
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class ConformityModuleResponse(BaseModel):
    """Conformity assessment module recommendation per CRA Art. 32."""
    module_id: str
    name: str
    description: str
    mandatory: bool
    alternatives: List[str]
    rationale: str


class ClassificationResponse(BaseModel):
    """Classification result based on core functionality."""
    classification: str
    category_id: Optional[str] = None
    category_name: str
    conformity_assessment: str
    conformity_module: ConformityModuleResponse
    compliance_deadline: str
    reporting_deadline: str
    cost_estimate_min: int
    cost_estimate_max: int
    automotive_exception: bool
    rationale: str


class CraAssessmentResponse(BaseModel):
    """Full CRA assessment response"""
    id: str
    product_id: str
    product_name: Optional[str] = None
    classification: Optional[str] = None
    classification_answers: Dict[str, Any] = {}
    product_type: str
    compliance_path: str = "direct_patch"
    compliance_deadline: Optional[str] = None
    assessment_date: Optional[datetime] = None
    assessor_id: Optional[str] = None
    status: str
    overall_compliance_pct: int = 0
    support_period_years: Optional[int] = None
    support_period_justification: Optional[str] = None
    support_period_end: Optional[str] = None
    eoss_date: Optional[str] = None
    notes: Optional[str] = None
    automotive_exception: bool = False
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    requirement_statuses: List[CraRequirementStatusResponse] = []
    compensating_controls: List[CompensatingControlResponse] = []

    class Config:
        from_attributes = True


class CraAssessmentListItem(BaseModel):
    """Lightweight assessment for list views"""
    id: str
    product_id: str
    product_name: Optional[str] = None
    classification: Optional[str] = None
    product_type: str
    status: str
    overall_compliance_pct: int = 0
    compliance_deadline: Optional[str] = None
    updated_at: Optional[datetime] = None


class CraAssessmentListResponse(BaseModel):
    """List of assessments"""
    assessments: List[CraAssessmentListItem]
    total: int


class AutoMapResponse(BaseModel):
    """Result of auto-mapping TARA artifacts to CRA requirements"""
    mapped_count: int
    mappings: List[Dict[str, Any]]


class TargetMarketEnum(str, Enum):
    """Target market for inventory items"""
    EU = "eu"
    NON_EU = "non_eu"
    GLOBAL = "global"


class InventoryItemCreate(BaseModel):
    """Create an inventory item for CRA assessment"""
    assessment_id: str
    sku: str
    firmware_version: Optional[str] = None
    units_in_stock: int = 0
    units_in_field: int = 0
    oem_customer: Optional[str] = None
    target_market: TargetMarketEnum = TargetMarketEnum.EU
    last_production_date: Optional[str] = None
    notes: Optional[str] = None


class InventoryItemUpdate(BaseModel):
    """Update an inventory item"""
    sku: Optional[str] = None
    firmware_version: Optional[str] = None
    units_in_stock: Optional[int] = None
    units_in_field: Optional[int] = None
    oem_customer: Optional[str] = None
    target_market: Optional[TargetMarketEnum] = None
    last_production_date: Optional[str] = None
    notes: Optional[str] = None


class InventoryItemResponse(BaseModel):
    """Inventory item response"""
    id: str
    assessment_id: str
    sku: str
    firmware_version: Optional[str] = None
    units_in_stock: int = 0
    units_in_field: int = 0
    oem_customer: Optional[str] = None
    target_market: str = "eu"
    last_production_date: Optional[str] = None
    notes: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class InventorySummary(BaseModel):
    """Summary of inventory for an assessment"""
    total_skus: int
    total_units_in_stock: int
    total_units_in_field: int
    eu_units: int
    non_eu_units: int
    oems: List[str]
